Add network support for WinDbg/KD (KDNET) ##debug

abcSup opened the pull request:
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

* Add network support for WinDbg/KD (KDNET).
* Add Windows 10 (May 2019 Update) build 18362 profile
* Add iob_net io_backend plugin for shlr/windbg
* Change r_socket_connect to bind source port for UDP connection

**Test plan**

```
# windbg://<remote-ip>:<port>:<key>
$ r2 -d windbg://192.168.112.144:49152:1.1.1.1
```

**Related issues**

#17130

trufae commented on libr/socket/socket.c:
```diff
@@ -301,16 +301,40 @@ R_API bool r_socket_connect(RSocket *s, const char *host, const char *port, int
+				s->sa.sin_port = htons (s->port); // TODO honor etc/services
+				if (bind (s->fd, (struct sockaddr *)&s->sa, sizeof (s->sa)) < 0) {
+					r_sys_perror ("bind");
+#ifdef _MSC_VER
```
why not use `__WINDOWS__` ?

abcSup replied:
No reason as I followed some code in `r_socket_listen`. I can change it.

trufae commented on libr/socket/socket.c:
```diff
@@ -469,8 +493,14 @@ R_API bool r_socket_listen(RSocket *s, const char *port, const char *certfile) {
 #endif
-	if ((s->fd = socket (AF_INET, SOCK_STREAM, IPPROTO_TCP)) == R_INVALID_SOCKET) {
-		return false;
+	if (s->proto == R_SOCKET_PROTO_UDP) {
```
better use switch(s->proto) { because the "else" of UDP is not TCP, there's also UNIX sockets, and other kind of sockets may be supported in the future

trufae replied:
also may be better to write a function that converts an RSocketProto into posix IPPROTO_

abcSup replied:
R_SOCKET_PROTO_* is defined with IPPROTO_* in r_socket.h already. I will just use that instead.
https://github.com/radareorg/radare2/blob/ab7a4b804545a80f496ccf93228d1d14f86adbbb/libr/include/r_socket.h#L86-L87

trufae commented on shlr/windbg/iob_net.c:
```diff
@@ -0,0 +1,470 @@
+#include  <r_crypto.h>
```
add your copyright line

trufae commented on shlr/windbg/iob_net.c:
```diff
@@ -0,0 +1,470 @@
+		return NULL;
+	}
+
+	dkey = calloc (1, blocksz);
```
define and assign in the same line, you can save one line by removing the `ut8 *dkey, *key ,... from the begining of this function`

trufae commented on shlr/windbg/iob_net.c:
```diff
@@ -0,0 +1,470 @@
+}
+
+static ut8 *_hmacsha256(ut8 *key, int ksize, ut8 *msg, int msize) {
+	size_t i, blocksz = 64;
```
this blocksize looks constant, make it `const`

XVilka replied:
By the way, the hmac function can be moved to r_crypto or r_hash, I think.

abcSup replied:
> By the way, the hmac function can be moved to r_crypto or r_hash, I think.

I thought about that, but I will need to look up on how to implement HMAC for other hashing algo first.

trufae commented on shlr/windbg/iob_net.c:
```diff
@@ -0,0 +1,470 @@
+		ikey[i] ^= 0x36;
+	}
+
+	ut8 *rdata = malloc (blocksz + msize);
```
check for null on every malloc

trufae commented on shlr/windbg/iob_net.c:
```diff
@@ -0,0 +1,470 @@
+	size_t i;
+	char *host = strdup (path);
+	char *port = strchr (host, ':');
+	if (!port || *port == '\0') {
```
use STR_IS_EMPTY

trufae replied:
it leaks host

trufae commented on shlr/windbg/iob_net.c:
```diff
@@ -0,0 +1,470 @@
+	return (void *) obj;
+}
+
+static int iob_net_close(void *p) {
```
make it return bool

trufae commented on shlr/windbg/iob_net.c:
```diff
@@ -0,0 +1,470 @@
+	return ret;
+}
+
+static int _encrypt(iobnet_t *obj, ut8 *buf, int size, int type) {
```
return bool and handle true/false, rtight now it's always returning -1

trufae commented on shlr/windbg/iob_net.c:
```diff
@@ -0,0 +1,470 @@
+static int _encrypt(iobnet_t *obj, ut8 *buf, int size, int type) {
+	int ret = -1;
+
+	RCrypto *cry = r_crypto_new ();
```
chk for nul

trufae commented on shlr/windbg/kd.h:
```diff
@@ -79,8 +79,9 @@ enum KD_PACKET_MANIPULATE_TYPE {
 
-#define KD_PACKET_DATA 0x30303030
-#define KD_PACKET_CTRL 0x69696969
+#define KD_PACKET_UNUSED	0x00000000
```
do not indent the values just use a single space

trufae commented on shlr/windbg/kd.h:
```diff
@@ -213,14 +226,14 @@ ct_assert(sizeof(kd_packet_t)==16);
 
-int kd_packet_is_valid (const kd_packet_t *p);
-int kd_packet_is_ack (const kd_packet_t *p);
+int kd_packet_is_valid(const kd_packet_t *p);
```
should return bool

XVilka commented:
Missing dependency with meson (https://github.com/radareorg/radare2/pull/17340/checks?check_run_id=904669209#step:9:861):
```
Undefined symbols for architecture x86_64:
  "_r_crypto_final", referenced from:
      __decrypt in libr_windbg.a(iob_net.o)
      __encrypt in libr_windbg.a(iob_net.o)
  "_r_crypto_free", referenced from:
      __decrypt in libr_windbg.a(iob_net.o)
      __encrypt in libr_windbg.a(iob_net.o)
  "_r_crypto_get_output", referenced from:
      __decrypt in libr_windbg.a(iob_net.o)
      __encrypt in libr_windbg.a(iob_net.o)
  "_r_crypto_new", referenced from:
      __decrypt in libr_windbg.a(iob_net.o)
      __encrypt in libr_windbg.a(iob_net.o)
  "_r_crypto_set_iv", referenced from:
      __decrypt in libr_windbg.a(iob_net.o)
      __encrypt in libr_windbg.a(iob_net.o)
  "_r_crypto_set_key", referenced from:
      __decrypt in libr_windbg.a(iob_net.o)
      __encrypt in libr_windbg.a(iob_net.o)
  "_r_crypto_use", referenced from:
      __decrypt in libr_windbg.a(iob_net.o)
      __encrypt in libr_windbg.a(iob_net
...
```

yossizap commented on shlr/windbg/windbg.c:
```diff
@@ -820,12 +827,19 @@ int windbg_sync(WindCtx *ctx) {
 
 	windbg_lock_enter (ctx);
 
+	ret = windbg_wait_packet (ctx, KD_PACKET_TYPE_UNUSED, NULL);
+	if (ret != KD_E_OK) {
+		ret = 0;
+		goto end;
+	}
```
Pipe mode freezes here with:

`Opened pipe /tmp/windbg.pipe with fd 0x6`

Seeing that KD_PACKET_TYPE_UNUSED wasn't expected in the previous pipe implementation, this should have been seen as something that needs testing / limiting to net mode.

GustavoLCR replied:
Tried on windows 10 connecting to a windows 8.1 after manually 'fixing' the static symbol problem, hanged on this as well

yossizap replied:
Seems to work now with Windows XP, can you confirm for 8.1 @GustavoLCR?

yossizap commented on shlr/windbg/windbg.c:
```diff
@@ -839,6 +853,7 @@ int windbg_sync(WindCtx *ctx) {
 		ret = 0;
 		goto end;
 	}
+*/
```
Don't leave it simply commented out if you want this merged. Have you figured out why this was done before? Pipe debug is still broken so this is hard for me to verify.

yossizap commented on shlr/windbg/kd.h:
```diff
@@ -203,6 +204,31 @@ typedef struct kd_packet_t {
+	ut8 type; // Channel Type - 0 Data, 1 Control
+}) kdnet_packet_t;
+
+// kdnet_data mask
+// 0xFFFFFF00 seqno | 0x80 direction | 0x7F padding size
+// direction:
+// - 0 Debuggee -> Debugger
+// - 1 Debugger -> Debuggee
```
Can you put this in the previous comment? It is out of context here

yossizap commented on shlr/windbg/kd.h:
```diff
@@ -203,6 +204,31 @@ typedef struct kd_packet_t {
+#define KDNET_DATA_SIZE 8
+#define KDNET_HMAC_SIZE 16
+
+// kdnet packet format
+// kdnet_packet_t
+// kdnet_data - encrypted
+// kd_packet - encrypted
+// kdnet_hmac
```
A high level review would help the reader understand what's going on here. Or at least a link to some documentation on these fields and what's supposed to be in them.

XVilka commented:
@GustavoLCR could you please take a look from the point of view of using radare2 from Windows for this feature?

abcSup commented:
I am sorry that my PR broke the pipe interface and will push an update soon as I include some changes to refactor the hmac function and add more documentation.

abcSup commented:
Looking into the linking issue on other platforms.

abcSup commented on shlr/windbg/transport.c:
```diff
@@ -4,22 +4,28 @@
 	NULL,
 };
 
 static io_backend_t *sel_backend = &iob_pipe;
```
Do you mind elaborating more on this? Do you mean that there is another static `io_backend_t ` in `RDebug` that keeps track of which interface to use and is separate from the one in `RIO`?

GustavoLCR replied:
transport.c gets linked in both rdebug and rio, so each gets its own instance of the variable

abcSup commented:
- Fixed `make` linking issues by including `r_crypto` in both `RIO` and `RDebug`.
- Fixed the static symbol problem pointed out by @GustavoLCR by refactoring the io backend to work similarly as `RIODesc`.

abcSup commented:
> * Connection doesn't work with a proper key (not `1.1.1.1`). The base36 to ut64 conversion seems wrong.

I realized I used "01234567890ABCDEFGHIJKLMNOPQRSTUVWXYZ" symbols instead of "01234567890abcdefghijklmnopqrstuvwxyz" symbols for base36. That should be the problem.

> * You should probably ignore `KD_PACKET_TYPE_UNUSED` in `windbg_wait_packet()`.

Good idea.

> Somewhat unrelated but Windows 8.1 profile seems wrong, build number should be 9600 and sp 0. Can you fix that in this/another pr?

I can submit a new PR to fix this. Which build are you using specifically? I see there are 6.3.9600.17415 Update 1 and 6.3.9600.16384 RTM. Anyway, I could just add them all at once including x86 and x64 later.

abcSup commented:
> * I managed to get a heap corruption somewhere. I'm not sure how/where, better investigate anyways.

@GustavoLCR Could you let me know what commands you used to get to the corruption?

abcSup commented:
- Fixed non-working interrupt when the debugger is in iob_net_read.
- Fixed base36 decoding to use the correct symbols "01234567890abcdefghijklmnopqrstuvwxyz".
- Refactor `windbg_wait_packet` to ignore KD_PACKET_TYPE_UNUSED

abcSup commented:
> The base36 decoding still doesn't work. See this valid key for example: `19ywxtyac3zgl.q6x77vxx82sl.1i2uwoth4iwjy.1mkxg0whhs63s`

Sorry for the lack of testing previously. I have just updated the PR to fix that and tested it to support up to `UT64_MAX - 1` value (`"3w5e11264sgse"`) in the keys.

XVilka commented:
@GustavoLCR please take a look when you have time.

GustavoLCR commented on shlr/windbg/windbg.c:
```diff
@@ -256,11 +259,15 @@ int windbg_wait_packet(WindCtx *ctx, const uint32_t type, kd_packet_t **p) {
 		}
 
+		if (pkt->type == KD_PACKET_TYPE_UNUSED) {
+			continue;
```
This will exaust `retries` if theres many unused packets unread

XVilka replied:
@abcSup please don't forget about this comment.

abcSup replied:
thanks for the reminder. I have added code to increase `retries` for unused KD packet.

GustavoLCR commented:
> > * I managed to get a heap corruption somewhere. I'm not sure how/where, better investigate anyways.
> 
> @GustavoLCR Could you let me know what commands you used to get to the corruption?

Couldn't reproduce anymore

> Which build are you using specifically? I see there are 6.3.9600.17415 Update 1 and 6.3.9600.16384 RTM. Anyway, I could just add them all at once including x86 and x64 later.

My vm is running 6.3.9600.17196, probably is the same profile as 17415

XVilka commented:
Please don't forget to send a PR to https://github.com/radareorg/radare2book/blob/master/src/debugger/windbg.md

abcSup commented:
> Check for ret here:
> https://github.com/radareorg/radare2/blob/5fcb21963c702d9c0ce6f02c4a650710f5b01dd6/shlr/windbg/windbg.c#L856-L858

Resolved

abcSup commented:
Let me rebase this and everything should be ready.

XVilka commented:
@GustavoLCR please take a look again - if you are good with this, lets merge it.

abcSup commented:
Sorry for the forced push again. I removed other windbg files (windbg.h, windbg.mk) that were meant to use the new winkd name.

nemanjan00 commented:
This commit broke building for me

Apperently SHA256_BLOCK_LENGTH is not defined. 

Build environment I used is aur script for radare2-git
